frontend/src/app/projects/models/add-t4c-source/add-t4c-source.component.ts

```diff
@@ -0,0 +1,230 @@
+      },
+      [Validators.required, this.uniqueNameValidator.bind(this)]
+    ),
+  });
+
+  constructor(
+    public projectService: ProjectService,
+    public modelService: ModelService,
+    public t4cInstanceService: T4CInstanceService,
+    public t4cRepositoryService: T4CRepoService,
+    public t4cModelService: T4cModelService
+  ) {}
+
+  ngOnInit(): void {
+    this.t4cInstanceService.listInstances().subscribe((instances) => {
+      this._instances.next(instances);
+    });
+
+    this.form.controls.t4c_instance_id.valueChanges.subscribe(
+      (t4c_instance_id) => {
+        this.form.controls.t4c_repository_id.reset();
+        if (this.form.controls.t4c_instance_id.valid) {
+          this._repositories.next(undefined);
+          this.t4cRepositoryService
+            .getT4CRepositories(t4c_instance_id!)
+            .subscribe((repositories) => {
+              this._repositories.next(repositories);
+              this.form.controls.t4c_repository_id.enable();
+            });
+        } else {
+          this._repositories.next(undefined);
+          this.form.controls.t4c_repository_id.disable();
+        }
+      }
+    );
+
+    this.form.controls.t4c_repository_id.valueChanges.subscribe(
+      (t4c_repository_id) => {
+        this.form.controls.name.reset();
+        if (this.form.controls.t4c_repository_id.valid) {
+          this._models.next(undefined);
+          this.t4cModelService
+            .listT4CModelsOfRepository(
+              this.projectService.project!.slug,
+              this.modelService.model!.slug,
+              this.form.value.t4c_instance_id!,
+              t4c_repository_id!
+            )
+            .subscribe((models) => {
+              this._models.next(models);
+              this.form.controls.name.enable();
+            });
+        } else {
+          this._models.next([]);
+          this.form.controls.name.disable();
+        }
+      }
+    );
+
+    this.resetToInstance();
+  }
+
+  onEditing() {
+    this.editing = true;
+    this.form.enable({ emitEvent: false });
+  }
+
+  resetToInstance() {
+    this.editing = false;
+    this.t4cModelService._t4cModel
+      .pipe(filter(Boolean), take(1))
+      .subscribe((model: T4CModel) => {
+        this.form.controls.t4c_instance_id.patchValue(
+          model.repository.instance.id
+        );
+
+        this._repositories
+          .pipe(filter(Boolean), delay(50), take(1))
+          .subscribe(() => {
+            this.form.controls.t4c_repository_id.patchValue(
+              model.repository.id
+            );
+            this._models
+              .pipe(filter(Boolean), delay(50), take(1))
+              .subscribe(() => {
+                this.form.controls.name.patchValue(model.name);
+                this.form.disable({ emitEvent: false });
+              });
+          });
+      });
+  }
+
+  onSubmit() {
+    if (
+      this.form.valid &&
+      this.projectService.project &&
+      this.modelService.model
```

Comment:
I had also added these two lines to my "onSubmit()", but I think since the `component.html` has the `*ngIf=this.projectService.project && this.modelService.model` at the top, this is not really needed here.